skbio/stats/distance/_base.py

```diff
@@ -109,6 +109,57 @@ def __init__(self, data, ids=None):
         self._ids = ids
         self._id_index = self._index_list(self._ids)
 
+    @classonlymethod
+    @experimental(as_of="0.5.0-dev")
+    def from_iterable(cls, iterable, metric, key=None, keys=None):
+        """Create DissimilarityMatrix from an iterable given a metric.
+
+        Parameters
+        ----------
+        iterable : iterable
+            Iterable containing objects to compute pairwise dissimilarities on.
+        metric : callable
+            A function that takes two arguments and returns a float
+            representing the dissimilarity between the two arguments.
+        key : callable or metadata key, optional
+            A function that takes one argument and returns a string
+            representing the id of the element in the dissimilarity matrix.
+            Alternatively, a key to a `metadata` property if it exists for
+            each element in the `iterable`. If None, then default ids will be
+            used.
+        keys : iterable, optional
+            An iterable of the same length as `iterable`. Each element will be
+            used as the respective key.
+
+        Returns
+        -------
+        DissimilarityMatrix
+            The `metric` applied to all pairwise elements in the `iterable`.
+
+        Raises
+        ------
+        ValueError
+            If `key` and `keys` are both provided.
+
+        """
+        iterable = list(iterable)
+        if key is not None and keys is not None:
+            raise ValueError("Cannot use both `key` and `keys` at the same"
+                             " time.")
+
+        keys_ = None
+        if key is not None:
+            keys_ = [resolve_key(e, key) for e in iterable]
+        elif keys is not None:
+            keys_ = keys
+
+        dm = np.zeros((len(iterable),) * 2)
```

Comment:
Now that the entire matrix is filled in it's a bit faster to use `np.empty` instead of `np.zeros`. Can you update this call?